Add spec to the sap module (#12)
ballerina-platform/module-ballerinax-sap: docs/spec/spec.md

To initialize the client, you need a valid SAP URL and optional configuration parameters. Here's an example of how to create a new SAP client:

```ballerina
sap:Client sapClient = check new (string `https://${hostname}/sap/opu/odata/sap/API_SALES_ORDER_SRV`, {
  auth: {
    username,
    password
  }
});
```

#### 2.1.1 Security

The `sap:Client` provides secure HTTP methods for interacting with SAP's HTTP endpoints.

```ballerina
sap:Client sapClient = check new (string `https://${hostname}/sap/opu/odata/sap/API_SALES_ORDER_SRV`, {
  auth: {
    username: username,
    password: password
  },
  secureSocket: {
    cert: {
      path: truststore_path,
      password: "ballerina"
    }
  }
});
```

This client is equipped with built-in CSRF token authentication, ensuring compliance with SAP system security standards.

#### 2.1.2 Cookie

As per SAP's CSRF token authentication requirements, cookie functionality is always enabled.

### 2.2. Client action

The SAP client contains separate remote method representing each HTTP method such as `get`, `put`, `post`,
`delete`,`patch`,`head`,`options` and some custom remote methods.

#### 2.2.1 Entity body methods
 

> POST, PUT, DELETE, PATCH methods are considered as entity body methods. These remote methods contains RequestMessage

Reviewer: ```suggestion
POST, PUT, DELETE, and PATCH methods are considered as entity body methods. These remote methods contain RequestMessage
```

```

```ballerina
# The post() function can be used to send HTTP POST requests to HTTP endpoints.
remote isolated function post(string path, RequestMessage message, map<string|string[]>? headers = (),
        string? mediaType = (), TargetType targetType = <>)
        returns targetType|ClientError;

# The put() function can be used to send HTTP PUT requests to HTTP endpoints.
remote isolated function put(string path, RequestMessage message, map<string|string[]>? headers = (),
        string? mediaType = (), TargetType targetType = <>)
        returns targetType|ClientError;

# The patch() function can be used to send HTTP PATCH requests to HTTP endpoints.
remote isolated function patch(string path, RequestMessage message, map<string|string[]>? headers = (),
        string? mediaType = (), TargetType targetType = <>)
        returns targetType|ClientError;

# The delete() function can be used to send HTTP DELETE requests to HTTP endpoints.
remote isolated function delete(string path, RequestMessage message = (), map<string|string[]>? headers = (),
        string? mediaType = (), TargetType targetType = <>)
        returns targetType|ClientError;
```

> #### 2.2.2 Non Entity body methods

Reviewer: ```suggestion
#### 2.2.2 Non-Entity body methods
```

GET, HEAD, OPTIONS methods are considered as non entity body methods. These remote methods do not contain 

> RequestMessage, but the header map an optional param.

Reviewer: ```suggestion
RequestMessage, but the header map is an optional param.
```

    auth: {
        username,
        password
    }
});

json salesOrder = check sapClient->/A_SalesOrder/${salesOrder}();

// Same as the following :
json salesOrder = check sapClient->get(string `/A_SalesOrder/${salesOrder}`);
```

* Query parameter

A query parameter is passed as a key-value pair in the resource method call.
The supported types are `string`, `int`, `float`, `boolean`, `decimal`, and the `array` types of the aforementioned types.
The query param type can be nil as well.
```ballerina
// Making a GET request
json salesOrder = check sapClient->/A_SalesOrder/${salesOrder}(limit = 2);

// Same as the following :
json salesOrder = check sapClient->get(string `/A_SalesOrder/${salesOrder}?limit=2`);
```

### 2.3. Client action return types

The SAP client remote method supports the contextually expected return types. The client operation is able to 
infer the expected payload type from the LHS variable type. This is called as client payload binding support where the 
inbound response payload is accessed and parse to the expected type in the method signature. It is easy to access the

> payload directly rather manipulation `http:Response` using its support methods such as `getTextPayload()`, ..etc.

Reviewer: ```suggestion
payload directly, rather than manipulating `http:Response` using its support methods such as `getTextPayload()`, ..etc.
```
